core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -346,14 +346,14 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
 
     // Since the state change was successfully written to the log, unset the flag for a failed epoch fence
     hasFailedEpochFence = false
-    val (updatedProducerId, updatedProducerEpoch) =
+    val (updatedProducerId, updatedProducerEpoch, updatedLastProducerEpoch) =
       // If we overflowed on epoch bump, we have to set it as the producer ID now the marker has been written.
       if (clientTransactionVersion.supportsEpochBump() && nextProducerId != RecordBatch.NO_PRODUCER_ID) {
-        (nextProducerId, 0.toShort)
+        (nextProducerId, 0.toShort, producerEpoch)
```

Comment:
Is this correct?  Wouldn't the producerEpoch be set to max value by this time?

Reply:
I thought this was correct since this will be the correct transition from the coordinator point of view.
Otherwise we will have max epoch - 1 as the last epoch. This was originally changed due to how we check for valid epochs.

I would have to double check, but I think this was also how the retry is checked

Reply:
Retry for this case is doesn't look at lastProducerEpoch, so either way is fine there. I think the big question is if we get a benefit from initProducerId by not updating the epoch here. The valid epoch check can be changed if we decide not to update between prepare and complete.

Let me know if I'm missing something

Reply:
Yeah and then it would be set as the lastProducerEpoch in the transit metadata right? Unless this is not the behavior we expect/want

Reply:
Artem's Message Offline 

> I think it would be more consistent to preserver lastProducerEpoch in the prepareComplete transition: the client will see the sequence, max-1, then 0, but the epoch will transition as max-1, max, 0 (the max value is not visible to the client but it'll be there as part of transition), so mathematically if lastEpoch is what the client saw last, then it should be max-1

Reply:
Made changes accordingly, also updated validProducerEpoch based on the different cases